src/hash.rs

```diff
@@ -0,0 +1,45 @@
+use bincode;
+use failure::Fail;
+use serde::{Deserialize, Serialize};
+use sha3::{digest::generic_array::transmute, Digest, Sha3_256};
+
+/// Type of hash commonly used within the library.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Hash(pub [u8; 32]);
+
+/// Hashing errors.
+#[derive(Debug, Fail)]
+pub enum Error {
+    /// A serialization error in `compute_hash`.
+    #[fail(display = "Serialization error: {}", _0)]
+    ComputeHashSerialize(bincode::Error),
+}
+
+/// Computes the hash of serializable data.
+pub fn compute_hash<B: Serialize>(b: &B) -> Result<Hash, Error> {
+    let mut hasher = Sha3_256::new();
+    let ser = bincode::serialize(b).map_err(Error::ComputeHashSerialize)?;
```

Comment:
Is bincode::serialize produces stable serialisation for any data type? If it isn't then this will produce different hashes for the same data.
It looks like bincode has such issue for HashMaps: https://github.com/servo/bincode/issues/230

Reply:
You are right! Will make sure no `HashMap`s are hashed this way. In fact, it seems to be a more general problem which makes hashing a `HashMap` non-deterministic.